### ISSUE TITLE
session_regenerate_id() error found. 

### ISSUE THREAD

**Issue by daisukeiip** [phpredis/phpredis#818]
Please check.
## Source code

session_start();
session_regenerate_id(true);
## Error message

Severity: 4096 --> session_regenerate_id(): Failed to create(read) session ID: rediscluster (path: seed[]=192.168.23.0:6379&amp;seed[]=192.168.23.1:6379&amp;seed[]=192.168.23.2:6379&amp;seed[]=192.168.23.3:6379&amp;seed[]=192.168.23.4:6379&amp;seed[]=192.168.23.5:6379&amp;timeout=3&amp;read_timeout=3&amp;failover=error) /usr/share/nginx/service_0/application/controllers/Test.php 20
## PHP settings
# > vi php.d/redis.ini

session.save_handler=rediscluster
session.save_path="seed[]=192.168.23.0:6379&seed[]=192.168.23.1:6379&seed[]=192.168.23.2:6379&seed[]=192.168.23.3:6379&seed[]=192.168.23.4:6379&seed[]=192.168.23.5:6379&timeout=3&read_timeout=3&failover=error"
## Environment
- CentOS 6.7 (64bit)
- PHP 7.0.6
- PHP-FPM 7.0.7
- Nginx-1.10.0
- Redis 3.0.7

**Comment by jmara** [comment]
Still the case with the most recent version of the module?

Cheers,
Jan

**Comment by arkhamvm** [comment]
@jmara, I'm getting same error too, tested on latest PHP7 branch version.
Profiling:
```
**17531 14:09:32.621020 sendto(16, "*2\r\n$3\r\nGET\r\n$57\r\nPHPREDIS_CLUSTER_SESSION:46ce3e17ca86e3408bd188a82f9fb833\r\n", 77, MSG_DONTWAIT, NULL, 0) = 77**
17531 14:09:32.621080 poll([{fd=16, events=POLLIN|POLLPRI|POLLERR|POLLHUP}], 1, 0) = 1 ([{fd=16, revents=POLLIN}])
17531 14:09:32.621115 recvfrom(16, "$", 1, MSG_PEEK, NULL, NULL) = 1
17531 14:09:32.621147 poll([{fd=16, events=POLLIN|POLLERR|POLLHUP}], 1, 60000) = 1 ([{fd=16, revents=POLLIN}])
17531 14:09:32.621181 recvfrom(16, "$-1\r\n", 8192, MSG_DONTWAIT, NULL, NULL) = 5
17531 14:09:32.621216 poll([{fd=16, events=POLLIN|POLLPRI|POLLERR|POLLHUP}], 1, 0) = 0 (Timeout)
17531 14:09:32.621285 chdir("/home/web/cgi-bin") = 0
**17531 14:09:32.621335 stat("/home/web/my/php/project/ErrorException.php", {st_mode=S_IFREG|0664, st_size=3798, ...}) = 0**
```
Env mostly the same (ubuntu instead centos).

**Comment by fperske** [comment]
same error with
- phpredis 3.0.0
- php-fpm 7.0.12
- redis 3.2.4 in cluster-mode:

`Catchable fatal error: session_regenerate_id(): Failed to create(read) session ID: rediscluster (path: seed[]=192.168.108.241:6379&seed[]=192.168.108.242:6379&seed[]=192.168.108.243:6379&timeout=2&read_timeout=2&failover=error&persistent=1) in test.php on line 3`

with test.php:
```
<?php
session_start();
session_regenerate_id(true);
?>
```

and php.ini:
```
session.save_handler = rediscluster
session.save_path = "seed[]=192.168.108.241:6379&seed[]=192.168.108.242:6379&seed[]=192.168.108.243:6379&timeout=2&read_timeout=2&failover=error&persistent=1"
```

**Comment by fperske** [comment]
Error also occurs with phpmyadmin (latest 4.6.4). same environment as above

```
Catchable Fatal Error in ./libraries/session.lib.php#20
session_regenerate_id(): Failed to create(read) session ID: rediscluster (path: seed[]=192.168.108.241:6379&amp;seed[]=192.168.108.242:6379&amp;seed[]=192.168.108.243:6379&amp;timeout=2&amp;read_timeout=2&amp;failover=error&amp;persistent=1)

Backtrace

./libraries/session.lib.php#20: session_regenerate_id(boolean true)
./libraries/common.inc.php#733: PMA_secureSession()
./index.php#13: require_once(./libraries/common.inc.php)
```

Phpredis is therefore not suitable for productive use. Can you please correct this issue?

**Comment by ephramm** [comment]
Hi,

I am having the same issue.

- phpredis 3.0.0
- libapache2-mod-php 7.0.12
- redis 3.2.1 in cluster mode

Thanks for looking into this.

**Comment by dmytro-pe** [comment]
Hello.

I have same issue with:

- phpredis 3.0.0
- redis 3.2.5 in cluster mode
- php 7.0.13

Thank you very much! session_regenerate_id(..) does not work correctly.

**Comment by leuchtdiode** [comment]
I'm getting the issue

`[16-Dec-2016 09:57:06 Europe/Berlin] PHP Catchable fatal error:  session_regenerate_id(): Failed to create(read) session ID: redis (path: tcp://[ip]:6379) in /appl/www/htdocs/cms/vendor/zendframework/zendframework1/library/Zend/Session.php on line 320`

on:

- Red Hat 7.3
- phpredis 3.1.0
- redis 3.2.6
- php 7.0.14

The weird thing is: I don't get the error on 3.0.0.

Any ideas?

Thx,
Alex

**Comment by inwel** [comment]
I had the same problem and found it to be a wrong variable type. The session handlers read method is expected to return a string. This workaround fixes it:

```
<?php
ini_set('display_errors', true);
class MySessionHandler extends SessionHandler
{
        public function read($session_id) {
                return (string)parent::read($session_id);
        }
}

$MySessionHandler = new MySessionHandler();
session_set_save_handler($MySessionHandler, true);

session_start();
var_dump(session_id());

session_regenerate_id(true);
var_dump(session_id());
?>
```
also see: https://bugs.php.net/bug.php?id=71187

**Comment by mahngiel** [comment]
Came across this as well with Docker `php:7.0-apache` + `redis-3.1.0` with Redis version 3.2.6.

Confirm workaround posted above works. Here's usage with PHP7's anonymous class 
```php
/**
 * @link https://github.com/phpredis/phpredis/issues/818
 * For PECL Redis 3.1.0,
 */
session_set_save_handler( new class extends \SessionHandler {
    public function read( $session_id )
    {
        return (string)parent::read( $session_id );
    }
}, true);
```

**Comment by jportoles** [comment]
I'm seeing this issue as well. I think that commit 18c7ab007649f068075880984e59938b849b5ab5 should fix it, but 3.1.1 is not published on PECL yet.

**Comment by madsi1m** [comment]
Has this been fixed yet?

**Comment by madsi1m** [comment]
Hi any updates on this?

**Comment by jportoles** [comment]
@madsi1m 3.1.1 is up on PECL (https://pecl.php.net/package/redis), it should resolve it if you have it installed properly.

**Comment by spsellars** [comment]
I was getting the same error even with 3.1.1, turned out to be related to [this](https://bugs.php.net/bug.php?id=74105) bug related to older kernels (which many OpenVZ VPS providers leave you stuck with.)  Recompiling PHP with the updated random.c file fixed it.

**Comment by madsi1m** [comment]
looks like it is working again. Thanks for the update

**Comment by daisukeiip** [comment]
PHP5.6 "session_regenerate_id(true)"  operate normally. but PHP7 dosen't.
session.save_handler = **rediscluster** 

[environment]
CentOS 7 + Redis 3.0.7 (Cluster)

$ yum list installed | grep -e php-pecl-redis -e nginx -e php-fpm -e php.x86_64
```
nginx.x86_64
php.x86_64                      7.1.8-1.el7.remi               @remi-php71      
php-fpm.x86_64                  7.1.8-1.el7.remi               @remi-php71      
php-pecl-redis.x86_64           3.1.2-1.el7.remi.7.1           @remi-php71 
```

$ cat /etc/php.d/50-redis.ini
```
...
session.save_handler = rediscluster
session.save_path = "seed[]=192.168.23.0:6379&seed[]=192.168.23.1:6379&seed[]=192.168.23.2:6379&seed[]=192.168.23.3:6379&seed[]=192.168.23.4:6379&seed[]=192.168.23.5:6379&timeout=3&read_timeout=3&failover=error&persistent=1"
...
```

$ cat /usr/share/nginx/html/index.php
```
<?php
session_start();
echo session_id();	// nvc43occpeecjkt32pmoivodpn
session_regenerate_id(true);
echo session_id();	// nvc43occpeecjkt32pmoivodpn (dosen't change)

```

**Comment by michael-grunder** [comment]
Thanks I'll take a look.

**Comment by yatsukhnenko** [comment]
#1199

**Comment by yatsukhnenko** [comment]
Fixed in >= 3.1.3
